<a href="https://colab.research.google.com/github/indahputri24/UTS_Business_Intelligence_Kelompok_11/blob/main/ETL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Extract**

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
# Membaca data
gc = pd.read_csv("/content/GroceryDataset.csv")

In [ ]:
gc.head()

,Sub Category,Price,Discount,Rating,Title,Currency,Feature,Product Description
0,Bakery & Desserts,$56.99,No Discount,Rated 4.3 out of 5 stars based on 265 reviews.,"David’s Cookies Mile High Peanut Butter Cake, ...",$,"""10"""" Peanut Butter Cake\nCertified Kosher OU-...",A cake the dessert epicure will die for!Our To...
1,Bakery & Desserts,$159.99,No Discount,Rated 5 out of 5 stars based on 1 reviews.,"The Cake Bake Shop 8"" Round Carrot Cake (16-22...",$,Spiced Carrot Cake with Cream Cheese Frosting ...,"Due to the perishable nature of this item, ord..."
2,Bakery & Desserts,$44.99,No Discount,Rated 4.1 out of 5 stars based on 441 reviews.,"St Michel Madeleine, Classic French Sponge Cak...",$,100 count\nIndividually wrapped\nMade in and I...,Moist and buttery sponge cakes with the tradit...
3,Bakery & Desserts,$39.99,No Discount,Rated 4.7 out of 5 stars based on 9459 reviews.,"David's Cookies Butter Pecan Meltaways 32 oz, ...",$,Butter Pecan Meltaways\n32 oz 2-Pack\nNo Prese...,These delectable butter pecan meltaways are th...
4,Bakery & Desserts,$59.99,No Discount,Rated 4.5 out of 5 stars based on 758 reviews.,"David’s Cookies Premier Chocolate Cake, 7.2 lb...",$,"""10"" Four Layer Chocolate Cake\nCertified Kosh...",A cake the dessert epicure will die for!To the...


In [ ]:
gc.shape

(1757, 8)

### Menampilkan struktur kolom setiap data

In [ ]:
gc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1757 entries, 0 to 1756
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Sub Category         1757 non-null   object
 1   Price                1754 non-null   object
 2   Discount             1757 non-null   object
 3   Rating               682 non-null    object
 4   Title                1757 non-null   object
 5   Currency             1752 non-null   object
 6   Feature              1739 non-null   object
 7   Product Description  1715 non-null   object
dtypes: object(8)
memory usage: 109.9+ KB


# **Transform**

## **Data Cleansing**

### Pengecekkan kolom Sub Category

In [ ]:
# Mengecek nilai untuk kolom price
gc["Sub Category"].unique()[:20]

array(['Bakery & Desserts', 'Beverages & Water', 'Breakfast', 'Candy',
       'Cleaning Supplies', 'Coffee', 'Deli', 'Floral', 'Gift Baskets',
       'Household', 'Kirkland Signature Grocery',
       'Laundry Detergent & Supplies', 'Meat & Seafood', 'Organic',
       'Pantry & Dry Goods', 'Paper & Plastic Products', 'Poultry',
       'Seafood', 'Snacks'], dtype=object)

### Pengecekkan kolom price, discount, dan rating

In [ ]:
# Mengecek nilai untuk kolom price
gc["Price"].unique()[:10]

array(['$56.99 ', '$159.99 ', '$44.99 ', '$39.99 ', '$59.99 ', '$74.99 ',
       '$29.99 ', '$62.99 ', '$299.99 ', '$54.99 '], dtype=object)

Setelah melakukan pengecekkan nilai unik pada kolom `Price` terdapat simbol unik berupa dolar ($) sehingga membuat tipe data kolom tersebut menjadi kategorik

In [ ]:
# Mengecek nilai untuk kolom discount
gc["Discount"].unique()[:10]

array(['No Discount', 'After $30 OFF', 'After $5 OFF',
       'This item is not returnable.', 'After $20 OFF', 'After $4.50 OFF',
       'After $4 OFF', 'After $2.80 OFF', 'After $8 OFF',
       'After $3.60 OFF'], dtype=object)

Setelah melakukan pengecekkan nilai unik pada kolom `Discount` terdapat simbol unik berupa dolar ($) dan beberapa teks sehingga membuat tipe data kolom tersebut menjadi kategorik

In [ ]:
# Mengecek nilai untuk kolom rating
gc["Rating"].unique()[:10]

array(['Rated 4.3 out of 5 stars based on 265 reviews.',
       'Rated 5 out of 5 stars based on 1 reviews.',
       'Rated 4.1 out of 5 stars based on 441 reviews.',
       'Rated 4.7 out of 5 stars based on 9459 reviews.',
       'Rated 4.5 out of 5 stars based on 758 reviews.',
       'Rated 4.4 out of 5 stars based on 369 reviews.',
       'Rated 4.7 out of 5 stars based on 2241 reviews.',
       'Rated 4.4 out of 5 stars based on 232 reviews.',
       'Rated 4.4 out of 5 stars based on 1679 reviews.',
       'Rated 5 out of 5 stars based on 2 reviews.'], dtype=object)

Setelah melakukan pengecekkan nilai unik pada kolom `Rating` terdapat penggabungan antara rating dengan jumlah reviews sehingga membuat tipe data kolom tersebut menjadi kategorik

### Pengecekkan kolom title, currency, feature, dan product description

In [ ]:
# Mengecek kolom title
gc["Title"].head(10)

,Title
0,"David’s Cookies Mile High Peanut Butter Cake, ..."
1,"The Cake Bake Shop 8"" Round Carrot Cake (16-22..."
2,"St Michel Madeleine, Classic French Sponge Cak..."
3,"David's Cookies Butter Pecan Meltaways 32 oz, ..."
4,"David’s Cookies Premier Chocolate Cake, 7.2 lb..."
5,David's Cookies Mango & Strawberry Cheesecake ...
6,"La Grande Galette French Butter Cookies, 1.3 l..."
7,David's Cookies No Sugar Added Cheesecake & Ma...
8,David's Cookies Brownie and Cookie Combo Pack
9,"The Cake Bake Shop 8"" Round Chocolate Cake (16..."


In [ ]:
# Mengecek nilai unik pada kolom currency
gc["Currency"].unique()

array(['$', nan], dtype=object)

Pada kolom `Currency` terdapat dua nilai unik, yaitu $ dan nan (nilai kosong)

> Add blockquote



In [ ]:
# Mengecek nilai pada kolom feature
gc["Feature"].head(5)

,Feature
0,"""10"""" Peanut Butter Cake\nCertified Kosher OU-..."
1,Spiced Carrot Cake with Cream Cheese Frosting ...
2,100 count\nIndividually wrapped\nMade in and I...
3,Butter Pecan Meltaways\n32 oz 2-Pack\nNo Prese...
4,"""10"" Four Layer Chocolate Cake\nCertified Kosh..."


Pada kolom `Feature` terdapat tanda kutip (") berlebih dan newline (\n)

In [ ]:
# Mengecek nilai
gc["Product Description"].head(5)

,Product Description
0,A cake the dessert epicure will die for!Our To...
1,"Due to the perishable nature of this item, ord..."
2,Moist and buttery sponge cakes with the tradit...
3,These delectable butter pecan meltaways are th...
4,A cake the dessert epicure will die for!To the...


In [ ]:
# Mengecek nilai unik pada kolom feature
gc["Product Description"].str.contains("\n", na=False).sum()

np.int64(893)

Pada kolom `Produk Description`  ditemukan newline (\n) yang digunakan untuk membuat baris baru

### Pengecekkan nilai kosong (missing value)

In [ ]:
# Menampilkan nilai kosong (missing value)
print("Jumlah data yang memiliki nilai kosong:")
print(gc.isnull().sum())

Jumlah data yang memiliki nilai kosong:
Sub Category              0
Price                     3
Discount                  0
Rating                 1075
Title                     0
Currency                  5
Feature                  18
Product Description      42
dtype: int64


Setelah melakukan pengecekkan nilai kosong, terdapat beberapa kolom yang memiliki missing value, yaitu:
- `Price`: 3
- `Rating`: 1075
- `Currency`: 5
- `Feature`: 18
- `Product Description`: 42

Nilai kosong pada kolom `Price` belum bisa dilakukan imputasi dengan nilai median  karena masih berupa teks dan mengandung simbol ($)

Nilai kosong pada kolom `Rating` belum bisa dilakukan imputasi karena masih berupa teks dan belum dipisahkan menjadi variabel numerik

#### Mengisi nilai kosong pada kolom Currency

In [ ]:
# Mengisi nilai kosong dengan USD
gc["Currency"] = gc["Currency"].fillna("USD")

#### Mengisi nilai kosong pada kolom Feature



In [ ]:
# Mengisi nilai kosong dengan No Feature
gc["Feature"] = gc["Feature"].fillna("No Feature")

#### Mengisi nilai kosong pada kolom Product Description

In [ ]:
# Mengisi nilai kosong dengan No Description
gc["Product Description"] = gc["Product Description"].fillna("No Description")

In [ ]:
print("Jumlah data yang memiliki nilai kosong:")
print(gc.isnull().sum())

Jumlah data yang memiliki nilai kosong:
Sub Category              0
Price                     3
Discount                  0
Rating                 1075
Title                     0
Currency                  0
Feature                   0
Product Description       0
dtype: int64


### Pengecekkan data duplikat

In [ ]:
print("Jumlah data duplikat:", gc.duplicated().sum())

Jumlah data duplikat: 3


In [ ]:
# Menampilkan data yang duplikat
print("Data yang duplikat:")
gc[gc.duplicated()]

Data yang duplikat:


,Sub Category,Price,Discount,Rating,Title,Currency,Feature,Product Description
1344,Paper & Plastic Products,$16.19,After $3.80 OFF,NaN,"Ziploc Seal Top Freezer Bag, Gallon, 38-count,...",$,Seal Top Bags 1-Gallon Freezer Bags 38 Bag...,4 - 38 Count Boxes\n152 Gallon Freezer Bags To...
1345,Paper & Plastic Products,$16.19,After $3.80 OFF,NaN,"Ziploc Seal Top Freezer Bag, Gallon, 38-count,...",$,Seal Top Bags 1-Gallon Freezer Bags 38 Bag...,4 - 38 Count Boxes\n152 Gallon Freezer Bags To...
1346,Paper & Plastic Products,$16.19,After $3.80 OFF,NaN,"Ziploc Seal Top Freezer Bag, Gallon, 38-count,...",$,Seal Top Bags 1-Gallon Freezer Bags 38 Bag...,4 - 38 Count Boxes\n152 Gallon Freezer Bags To...


Setelah melakukan pengecekkan data duplikat terdapat tiga data duplikat

In [ ]:
# Menghapus data duplikat
gc = gc.drop_duplicates()
gc[gc.duplicated()]

,Sub Category,Price,Discount,Rating,Title,Currency,Feature,Product Description


In [ ]:
# Menampilkan ringkasan informasi
gc.describe()

,Sub Category,Price,Discount,Rating,Title,Currency,Feature,Product Description
count,1754,1751,1754,682,1754,1754,1754,1754
unique,19,184,42,483,1484,2,1402,1436
top,Snacks,$14.99,No Discount,No Reviews,"Kirkland Signature, Organic Almond Beverage, V...",$,No Feature,No Description
freq,293,81,1626,61,3,1749,18,42


## **Data Standardization**

### Standardization kolom Sub Category

In [ ]:
# Menyamakan data menjadi huruf kapital
gc["Sub Category"] = gc["Sub Category"].str.strip()
gc["Sub Category"] = gc["Sub Category"].str.title()
gc['Sub Category'].head(5)

,Sub Category
0,Bakery & Desserts
1,Bakery & Desserts
2,Bakery & Desserts
3,Bakery & Desserts
4,Bakery & Desserts


### Standardization kolom Price

In [ ]:
# Menghapus simbol dolar ($) pada kolom Price
gc["Price"] = gc["Price"].str.replace("$", "", regex=False)
gc['Price'].unique()[:10]

array(['56.99 ', '159.99 ', '44.99 ', '39.99 ', '59.99 ', '74.99 ',
       '29.99 ', '62.99 ', '299.99 ', '54.99 '], dtype=object)

### Standardization Kolom Discount

In [ ]:
# Mengekstrak angka diskon pada kolom Discount
gc["Discount"] = gc["Discount"].replace({
    "No Discount": "0",
    "This item is not returnable.": "0"
})

gc["Discount"] = gc["Discount"].str.extract(r"(\d+\.?\d*)")
gc["Discount"].unique()[:10]

array(['0', '30', '5', '20', '4.50', '4', '2.80', '8', '3.60', '3'],
      dtype=object)

In [ ]:
gc[gc["Discount"].isna()]

,Sub Category,Price,Discount,Rating,Title,Currency,Feature,Product Description
1064,Meat & Seafood,NaN,NaN,Rated 4.5 out of 5 stars based on 4 reviews.,"Hillshire Farm Slow Roasted Turkey Breast, 11 ...",USD,Gluten Free No Artificial Ingredients No N...,Available at your local Costco warehouse or cl...
1414,Poultry,NaN,NaN,Rated 4.5 out of 5 stars based on 4 reviews.,"Hillshire Farm Slow Roasted Turkey Breast, 11 ...",USD,Gluten Free No Artificial Ingredients No N...,Available at your local Costco warehouse or cl...


### Standardization Kolom Rating

In [ ]:
# Mengubah nilai "No reviews" menjadi NaN
gc["Rating"] = gc["Rating"].replace("No Reviews", None)

# Memeriksa kembali nilai "No Reviews"
gc["Rating"].str.contains("No Reviews", na=False).sum()

np.int64(0)

### Standardization Kolom Title

In [ ]:
# Menghapus spasi ganda dan new line (\n)
gc["Title"] = gc["Title"].str.replace(r"\s+", " ", regex=True)
gc["Title"] = gc["Title"].str.replace("\n", " ", regex=False)
gc["Title"] = gc["Title"].str.strip()
gc["Title"].head()

,Title
0,"David’s Cookies Mile High Peanut Butter Cake, ..."
1,"The Cake Bake Shop 8"" Round Carrot Cake (16-22..."
2,"St Michel Madeleine, Classic French Sponge Cak..."
3,"David's Cookies Butter Pecan Meltaways 32 oz, ..."
4,"David’s Cookies Premier Chocolate Cake, 7.2 lb..."


### Standardization Kolom Currency

In [ ]:
# Mengubah simbol ($) menjadi USD
gc["Currency"] = gc["Currency"].replace("$", "USD")

### Standardization Kolom Feature

In [ ]:
# Menghapus new line (\n), spasi ganda, dan "
gc["Feature"] = gc["Feature"].str.replace("\n", " ", regex=False)
gc["Feature"] = gc["Feature"].str.replace('"', "", regex=False)
gc["Feature"] = gc["Feature"].str.replace(r"\s+", " ", regex=True)
gc["Feature"] = gc["Feature"].str.strip()
gc["Feature"].head()

,Feature
0,10 Peanut Butter Cake Certified Kosher OU-D 14...
1,Spiced Carrot Cake with Cream Cheese Frosting ...
2,100 count Individually wrapped Made in and Imp...
3,Butter Pecan Meltaways 32 oz 2-Pack No Preserv...
4,10 Four Layer Chocolate Cake Certified Kosher ...


### Standardization Kolom Product Description

In [ ]:
# Menghapus new line (\n) dan spasi ganda
gc["Product Description"] = gc["Product Description"].str.replace("\n", " ", regex=False)
gc["Product Description"] = gc["Product Description"].str.replace(r"\s+", " ", regex=True)
gc["Product Description"] = gc["Product Description"].str.strip()
gc["Product Description"].head()

,Product Description
0,A cake the dessert epicure will die for!Our To...
1,"Due to the perishable nature of this item, ord..."
2,Moist and buttery sponge cakes with the tradit...
3,These delectable butter pecan meltaways are th...
4,A cake the dessert epicure will die for!To the...


In [ ]:
gc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1754 entries, 0 to 1756
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Sub Category         1754 non-null   object
 1   Price                1751 non-null   object
 2   Discount             1752 non-null   object
 3   Rating               621 non-null    object
 4   Title                1754 non-null   object
 5   Currency             1754 non-null   object
 6   Feature              1754 non-null   object
 7   Product Description  1754 non-null   object
dtypes: object(8)
memory usage: 123.3+ KB


Dari pengecekkan tipe data terdapat kesalahan tipe data pada kolom `Price`, `Rating`,dan `Discount`yang seharusnya bertipe data numerik

## **Data Validation**

### Validasi kolom price

In [ ]:
# Mengubah tipe data menjadi numerik
gc["Price"] = pd.to_numeric(gc["Price"], errors="coerce")
gc["Price"].dtype

dtype('float64')

### Validasi kolom Discount

In [ ]:
# Mengubah tipe data menjadi numerik
gc["Discount"] = pd.to_numeric(gc["Discount"], errors="coerce")
gc["Discount"].dtype

dtype('float64')

### Validasi kolom Rating

In [ ]:
# Mengekstrak nilai rating
gc["Rating_Value"] = gc["Rating"].str.extract(r"Rated\s([\d\.]+)")

# Mengekstrak jumlah review
gc["Review_Count"] = gc["Rating"].str.extract(r"based on (\d+)")
print(gc[["Rating_Value", 'Review_Count']].head(5))
gc[["Rating_Value", 'Review_Count']].dtypes

  Rating_Value Review_Count
0          4.3          265
1            5            1
2          4.1          441
3          4.7         9459
4          4.5          758


,0
Rating_Value,object
Review_Count,object


In [ ]:
# Mengubah tipe data menjadi numerik
gc["Rating_Value"] = pd.to_numeric(gc["Rating_Value"], errors="coerce")
gc["Review_Count"] = pd.to_numeric(gc["Review_Count"], errors="coerce")
gc[["Rating_Value", 'Review_Count']].dtypes

,0
Rating_Value,float64
Review_Count,float64


In [ ]:
# Mengisi nilai kosong pada kolom Rating_Value dan Review_Count
gc["Review_Count"] = gc["Review_Count"].fillna(0)
gc["Rating_Value"] = gc["Rating_Value"].fillna(0)

# Mengubah tipe data Review_Count menjadi Integer
gc["Review_Count"] = gc["Review_Count"].astype(int)
gc[["Rating_Value", 'Review_Count']].dtypes

,0
Rating_Value,float64
Review_Count,int64


### Validasi kolom Sub Category

In [ ]:
# Menghitung jumlah data pada setiap kategori produk
gc["Sub Category"].value_counts()

,count
Sub Category,
Snacks,293
Pantry & Dry Goods,171
Candy,154
Beverages & Water,148
Meat & Seafood,144
Kirkland Signature Grocery,122
Coffee,95
Cleaning Supplies,94
Gift Baskets,89


### Validasi kolom Title

In [ ]:
# Mengecek apakah masih ada spasi ganda atau karakter new line (\n)
print("Jumlah yang mengandung spasi ganda:", gc["Title"].str.contains(r"\s{2,}", na=False).sum())

print("Jumlah yang mengandung new line:", gc["Title"].str.contains("\n", na=False).sum())


Jumlah yang mengandung spasi ganda: 0
Jumlah yang mengandung new line: 0


### Validasi kolom Currency

In [ ]:
# Menampilkan nilai unik dan tipe data
gc["Currency"].unique()

array(['USD'], dtype=object)

### Validasi kolom Feature dan Product Description

In [ ]:
# Kolom Feature
print("Kolom Feature")
print("Jumlah yang mengandung spasi ganda:", gc["Feature"].str.contains(r"\s{2,}", na=False).sum())
print("Jumlah yang mengandung new line:", gc["Feature"].str.contains("\n", na=False).sum())

# Kolom Product Description
print("\nKolom Product Description")
print("Jumlah yang mengandung spasi ganda:", gc["Product Description"].str.contains(r"\s{2,}", na=False).sum())
print("Jumlah yang mengandung new line:", gc["Product Description"].str.contains("\n", na=False).sum())



Kolom Feature
Jumlah yang mengandung spasi ganda: 0
Jumlah yang mengandung new line: 0

Kolom Product Description
Jumlah yang mengandung spasi ganda: 0
Jumlah yang mengandung new line: 0


### Verifikasi Hasil Data Validation dan Pengecekkan Nilai Kosong

In [ ]:
gc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1754 entries, 0 to 1756
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Sub Category         1754 non-null   object 
 1   Price                1732 non-null   float64
 2   Discount             1752 non-null   float64
 3   Rating               621 non-null    object 
 4   Title                1754 non-null   object 
 5   Currency             1754 non-null   object 
 6   Feature              1754 non-null   object 
 7   Product Description  1754 non-null   object 
 8   Rating_Value         1754 non-null   float64
 9   Review_Count         1754 non-null   int64  
dtypes: float64(3), int64(1), object(6)
memory usage: 150.7+ KB


In [ ]:
# Menampilkan kembali missing value
print(gc.isnull().sum())

Sub Category              0
Price                    22
Discount                  2
Rating                 1133
Title                     0
Currency                  0
Feature                   0
Product Description       0
Rating_Value              0
Review_Count              0
dtype: int64


#### Imputasi kolom Price

In [ ]:
# Mengisi nilai kosong dengan nilai median
gc["Price"] = gc["Price"].fillna(gc["Price"].median())

#### Imputasi kolom Discount

In [ ]:
gc["Discount"] = gc["Discount"].fillna(0)



#### Memeriksa data dan nilai kosong kembali (Missing Value)

In [ ]:
gc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1754 entries, 0 to 1756
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Sub Category         1754 non-null   object 
 1   Price                1754 non-null   float64
 2   Discount             1754 non-null   float64
 3   Rating               621 non-null    object 
 4   Title                1754 non-null   object 
 5   Currency             1754 non-null   object 
 6   Feature              1754 non-null   object 
 7   Product Description  1754 non-null   object 
 8   Rating_Value         1754 non-null   float64
 9   Review_Count         1754 non-null   int64  
dtypes: float64(3), int64(1), object(6)
memory usage: 150.7+ KB


In [ ]:
print(gc.isnull().sum())

Sub Category              0
Price                     0
Discount                  0
Rating                 1133
Title                     0
Currency                  0
Feature                   0
Product Description       0
Rating_Value              0
Review_Count              0
dtype: int64


## **Data Enrichment**

In [ ]:
# Membuat kolom Final_Price (harga setelah diskon)
gc["Final_Price"] = gc["Price"] - gc["Discount"]
gc["Final_Price"] = gc["Final_Price"].clip(lower=0)

In [ ]:
# Membuat kategori rating
gc["Rating_Category"] = pd.cut(
    gc["Rating_Value"],
    bins=[0, 2, 4, 5],
    labels=["Low", "Medium", "High"]
)

In [ ]:
# Membuat kolom Popularity
gc["Popularity"] = gc["Review_Count"]


In [ ]:
# Membuat kategori popularitas berdasarkan jumlah review
gc["Popularity_Category"] = pd.cut(
    gc["Review_Count"],
    bins=[0, 50, 200, 10000],
    labels=["Low", "Medium", "High"]
)


In [ ]:
# Verifikasi Hasil Data Enrichment
print(gc.head())

        Sub Category   Price  Discount  \
0  Bakery & Desserts   56.99       0.0   
1  Bakery & Desserts  159.99       0.0   
2  Bakery & Desserts   44.99       0.0   
3  Bakery & Desserts   39.99       0.0   
4  Bakery & Desserts   59.99       0.0   

                                            Rating  \
0   Rated 4.3 out of 5 stars based on 265 reviews.   
1       Rated 5 out of 5 stars based on 1 reviews.   
2   Rated 4.1 out of 5 stars based on 441 reviews.   
3  Rated 4.7 out of 5 stars based on 9459 reviews.   
4   Rated 4.5 out of 5 stars based on 758 reviews.   

                                               Title Currency  \
0  David’s Cookies Mile High Peanut Butter Cake, ...      USD   
1  The Cake Bake Shop 8" Round Carrot Cake (16-22...      USD   
2  St Michel Madeleine, Classic French Sponge Cak...      USD   
3  David's Cookies Butter Pecan Meltaways 32 oz, ...      USD   
4  David’s Cookies Premier Chocolate Cake, 7.2 lb...      USD   

                               

## **Data Filtering**

In [ ]:
# Menghapus harga tidak valid (<= 0)
gc = gc[gc["Price"] > 0]

In [ ]:
# Menghapus diskon tidak valid (< 0)
gc = gc[gc["Discount"] >= 0]

In [ ]:
# Menghapus rating tidak valid (> 5)
gc = gc[gc["Rating_Value"] <= 5]

In [ ]:
# Menghapus review count tidak valid (< 0)
gc = gc[gc["Review_Count"] >= 0]

In [ ]:
# Menghapus data tanpa nama produk
gc = gc[gc["Title"].notnull()]

In [ ]:
# Menghapus kategori kosong
gc = gc[gc["Sub Category"].notnull()]

In [ ]:
# Reset index setelah filtering
gc = gc.reset_index(drop=True)

### Verifikasi Hasil Data Filtering

In [ ]:
# Cek jumlah data setelah filtering
print("Jumlah data setelah filtering:", len(gc))

Jumlah data setelah filtering: 1754


In [ ]:
print(gc.head())

        Sub Category   Price  Discount  \
0  Bakery & Desserts   56.99       0.0   
1  Bakery & Desserts  159.99       0.0   
2  Bakery & Desserts   44.99       0.0   
3  Bakery & Desserts   39.99       0.0   
4  Bakery & Desserts   59.99       0.0   

                                            Rating  \
0   Rated 4.3 out of 5 stars based on 265 reviews.   
1       Rated 5 out of 5 stars based on 1 reviews.   
2   Rated 4.1 out of 5 stars based on 441 reviews.   
3  Rated 4.7 out of 5 stars based on 9459 reviews.   
4   Rated 4.5 out of 5 stars based on 758 reviews.   

                                               Title Currency  \
0  David’s Cookies Mile High Peanut Butter Cake, ...      USD   
1  The Cake Bake Shop 8" Round Carrot Cake (16-22...      USD   
2  St Michel Madeleine, Classic French Sponge Cak...      USD   
3  David's Cookies Butter Pecan Meltaways 32 oz, ...      USD   
4  David’s Cookies Premier Chocolate Cake, 7.2 lb...      USD   

                               

## **Data Aggregation**

#### Jumlah produk per kategori

In [ ]:
# Menghitung jumlah produk di setiap Sub Category
product_count = gc.groupby("Sub Category")["Title"].count()

print("Jumlah produk per kategori:")
print(product_count)

Jumlah produk per kategori:
Sub Category
Bakery & Desserts                33
Beverages & Water               148
Breakfast                        21
Candy                           154
Cleaning Supplies                94
Coffee                           95
Deli                             21
Floral                           75
Gift Baskets                     89
Household                        81
Kirkland Signature Grocery      122
Laundry Detergent & Supplies     39
Meat & Seafood                  144
Organic                          33
Pantry & Dry Goods              171
Paper & Plastic Products         85
Poultry                           9
Seafood                          47
Snacks                          293
Name: Title, dtype: int64


#### Rata-rata harga per kategori

In [ ]:
# Menghitung rata-rata harga produk di setiap kategori
avg_price = gc.groupby("Sub Category")["Price"].mean()

print("\nRata-rata harga per kategori:")
print(avg_price)


Rata-rata harga per kategori:
Sub Category
Bakery & Desserts                83.414242
Beverages & Water                36.703514
Breakfast                        13.304286
Candy                            25.069221
Cleaning Supplies                30.261277
Coffee                           32.880526
Deli                            211.085238
Floral                           61.043333
Gift Baskets                     81.079888
Household                        28.335679
Kirkland Signature Grocery       28.119508
Laundry Detergent & Supplies     33.895128
Meat & Seafood                  218.323194
Organic                          22.420303
Pantry & Dry Goods               29.103450
Paper & Plastic Products         25.667647
Poultry                         144.878889
Seafood                         195.521489
Snacks                           21.544266
Name: Price, dtype: float64


#### Rata-rata rating per kategori

In [ ]:
# Menghitung rata-rata rating produk di setiap kategori
avg_rating = gc.groupby("Sub Category")["Rating_Value"].mean()

print("\nRata-rata rating per kategori:")
print(avg_rating)


Rata-rata rating per kategori:
Sub Category
Bakery & Desserts               3.593939
Beverages & Water               0.710811
Breakfast                       0.452381
Candy                           0.519481
Cleaning Supplies               1.739362
Coffee                          1.758947
Deli                            4.261905
Floral                          3.164000
Gift Baskets                    3.748315
Household                       0.000000
Kirkland Signature Grocery      0.718033
Laundry Detergent & Supplies    1.830769
Meat & Seafood                  3.803472
Organic                         1.121212
Pantry & Dry Goods              1.159064
Paper & Plastic Products        0.683529
Poultry                         4.477778
Seafood                         3.642553
Snacks                          0.553584
Name: Rating_Value, dtype: float64


#### Total review per kategori

In [ ]:
# Mengetahui kategori mana yang paling populer
total_review = gc.groupby("Sub Category")["Review_Count"].sum()

print("\nTotal review per kategori:")
print(total_review)


Total review per kategori:
Sub Category
Bakery & Desserts                19790
Beverages & Water                11542
Breakfast                         1424
Candy                            13563
Cleaning Supplies                16824
Coffee                           35724
Deli                             10708
Floral                          141257
Gift Baskets                     43169
Household                            0
Kirkland Signature Grocery       46151
Laundry Detergent & Supplies      9385
Meat & Seafood                   29161
Organic                          11070
Pantry & Dry Goods               14611
Paper & Plastic Products          2773
Poultry                            751
Seafood                          10306
Snacks                           17564
Name: Review_Count, dtype: int64


#### Rata-rata diskon per kategori

In [ ]:
# Mengetahui kategori mana yang sering diskon
avg_discount = gc.groupby("Sub Category")["Discount"].mean()

print("\nRata-rata diskon per kategori:")
print(avg_discount)


Rata-rata diskon per kategori:
Sub Category
Bakery & Desserts               0.000000
Beverages & Water               0.537162
Breakfast                       0.895238
Candy                           0.248701
Cleaning Supplies               0.839362
Coffee                          0.547368
Deli                            6.190476
Floral                          0.000000
Gift Baskets                    1.685393
Household                       0.000000
Kirkland Signature Grocery      0.217213
Laundry Detergent & Supplies    1.423077
Meat & Seafood                  3.194444
Organic                         0.000000
Pantry & Dry Goods              0.422807
Paper & Plastic Products        0.383529
Poultry                         0.000000
Seafood                         2.765957
Snacks                          0.310580
Name: Discount, dtype: float64


#### Menggabungkan semua hasil agregasi ke satu tabel

In [ ]:
# Menggabungkan semua hasil agregasi ke satu tabel
summary = gc.groupby("Sub Category").agg({
    "Title": "count",
    "Price": "mean",
    "Rating_Value": "mean",
    "Review_Count": "sum",
    "Discount": "mean"
})

# Mengganti nama kolom agar lebih jelas
summary.columns = [
    "Total_Product",
    "Avg_Price",
    "Avg_Rating",
    "Total_Review",
    "Avg_Discount"
]

print("\nSUMMARY PER KATEGORI:")
print(summary)


SUMMARY PER KATEGORI:
                              Total_Product   Avg_Price  Avg_Rating  \
Sub Category                                                          
Bakery & Desserts                        33   83.414242    3.593939   
Beverages & Water                       148   36.703514    0.710811   
Breakfast                                21   13.304286    0.452381   
Candy                                   154   25.069221    0.519481   
Cleaning Supplies                        94   30.261277    1.739362   
Coffee                                   95   32.880526    1.758947   
Deli                                     21  211.085238    4.261905   
Floral                                   75   61.043333    3.164000   
Gift Baskets                             89   81.079888    3.748315   
Household                                81   28.335679    0.000000   
Kirkland Signature Grocery              122   28.119508    0.718033   
Laundry Detergent & Supplies             39   33.89512

# **Data Warehouse (Star Schema)**

### Membuat Dim_Category

In [ ]:
# Ambil kategori unik
dim_category = gc[["Sub Category"]].drop_duplicates().reset_index(drop=True)

# Tambahkan ID (primary key)
dim_category["Category_ID"] = dim_category.index + 1

# Urutkan kolom
dim_category = dim_category[["Category_ID", "Sub Category"]]

print(dim_category.head())

   Category_ID       Sub Category
0            1  Bakery & Desserts
1            2  Beverages & Water
2            3          Breakfast
3            4              Candy
4            5  Cleaning Supplies


### Membuat Dim_Product

In [ ]:
# Ambil informasi produk
dim_product = gc[[
    "Title",
    "Feature",
    "Product Description",
    "Currency"
]].drop_duplicates().reset_index(drop=True)

# Tambahkan Product_ID
dim_product["Product_ID"] = dim_product.index + 1

# Urutkan kolom
dim_product = dim_product[[
    "Product_ID",
    "Title",
    "Feature",
    "Product Description",
    "Currency"
]]

print(dim_product.head())

   Product_ID                                              Title  \
0           1  David’s Cookies Mile High Peanut Butter Cake, ...   
1           2  The Cake Bake Shop 8" Round Carrot Cake (16-22...   
2           3  St Michel Madeleine, Classic French Sponge Cak...   
3           4  David's Cookies Butter Pecan Meltaways 32 oz, ...   
4           5  David’s Cookies Premier Chocolate Cake, 7.2 lb...   

                                             Feature  \
0  10 Peanut Butter Cake Certified Kosher OU-D 14...   
1  Spiced Carrot Cake with Cream Cheese Frosting ...   
2  100 count Individually wrapped Made in and Imp...   
3  Butter Pecan Meltaways 32 oz 2-Pack No Preserv...   
4  10 Four Layer Chocolate Cake Certified Kosher ...   

                                 Product Description Currency  
0  A cake the dessert epicure will die for!Our To...      USD  
1  Due to the perishable nature of this item, ord...      USD  
2  Moist and buttery sponge cakes with the tradit...      USD 

In [ ]:
# Menghubungkan Category ke Product
gc = gc.merge(dim_category, on="Sub Category", how="left")

# Menghubungkan Product ke Data
gc = gc.merge(dim_product, on=["Title","Feature","Product Description","Currency"], how="left")

### Membuat Fact Table

In [ ]:
fact_table = gc[[
    "Product_ID",
    "Category_ID",
    "Price",
    "Discount",
    "Final_Price",
    "Rating_Value",
    "Review_Count"
]]

print(fact_table.head())

   Product_ID  Category_ID   Price  Discount  Final_Price  Rating_Value  \
0           1            1   56.99       0.0        56.99           4.3   
1           2            1  159.99       0.0       159.99           5.0   
2           3            1   44.99       0.0        44.99           4.1   
3           4            1   39.99       0.0        39.99           4.7   
4           5            1   59.99       0.0        59.99           4.5   

   Review_Count  
0           265  
1             1  
2           441  
3          9459  
4           758  


# **Load**

In [ ]:
# Load ke CSV
dim_category.to_csv("dim_category.csv", index=False)
dim_product.to_csv("dim_product.csv", index=False)
fact_table.to_csv("fact_table.csv", index=False)

print("Data berhasil disimpan ke file CSV!")

Data berhasil disimpan ke file CSV!


In [ ]:
# Membuat koneksi database
conn = sqlite3.connect("data_warehouse.db")

# Load ke database
dim_category.to_sql("dim_category", conn, if_exists="replace", index=False)
dim_product.to_sql("dim_product", conn, if_exists="replace", index=False)
fact_table.to_sql("fact_product", conn, if_exists="replace", index=False)

print("Data berhasil dimasukkan ke database!")

Data berhasil dimasukkan ke database!


In [ ]:
# Cek isi database
query = "SELECT * FROM fact_product LIMIT 5"
print(pd.read_sql(query, conn))

   Product_ID  Category_ID   Price  Discount  Final_Price  Rating_Value  \
0           1            1   56.99       0.0        56.99           4.3   
1           2            1  159.99       0.0       159.99           5.0   
2           3            1   44.99       0.0        44.99           4.1   
3           4            1   39.99       0.0        39.99           4.7   
4           5            1   59.99       0.0        59.99           4.5   

   Review_Count  
0           265  
1             1  
2           441  
3          9459  
4           758  
